### ISSUE TITLE
Suggest conversion to method when function couldn't be found

### ISSUE THREAD

**Issue by mikusp** [luna/luna-studio#787]
![unknown](https://user-images.githubusercontent.com/2052578/35614760-6cd561b0-0670-11e8-930d-07ec46934396.png)

In this case, I think that it would be beneficial to suggest converting function application to a method. Possibly with a clickable button, similarly to some IDEs that can automatically transform code after suggesting e.g. wrapping code in try/catch.

This way, even if user types `reverse hello`, and `reverse :: Text -> Text` is not defined, TC sees that `Text.reverse` method is defined and suggest it instead.

**Comment by kustosz** [comment]
That would be interesting indeed. In fact, our TC should be a much more helpful beast in many aspects and I'm all for that. However, for now I think looking for ways to fix the error is something the TC should after it gets quickly to the errors in the first place. I.e. I'd say postpone until it's blazing fast.
For the general idea though, I'm 100% sold!

**Comment by wdanilo** [comment]
It's very interesting idea! In fact we've been already talking about this problem a lot with @kustosz. I'm not yet sold, there is another proposal for solving this issue, rather radical, however we need to talk about it. Namely, nodes in the visual form do not benefit very much from having "self" port. Maybe it would be much more clear for users to connect just to first port (it could be displayed in a special way, whatever). Anyway, the radical proposal is to make `foo.bar baz` always equivalent to `bar foo baz`. Such sugar syntax transformation can be implemented in such way, that it does not affect how we resolve types and compile programs and would allow connecting nodes in the visual editor always the same way. It's just an idea. I'm not sold to it, neither I'm sold to the idea of having a button for it. It needs further conversation, however, the problem you describe @mikusp is indeed very important.

**Comment by kustosz** [comment]
Apart from redesigning the whole language, @mikusp's idea is a broader one. What I'm seeing here is a kind of clippy-the-assistant thing (probably not that personal though :( ). Things like "the method `reevrse` does not exist, did you mean `reverse`" with a "yes, let's fix it", etc. Basically going out of our way to auto-fix such obvious mistakes, which will happen in different shapes no matter the language semantics.

**Comment by wdanilo** [comment]
@kustosz "redesigning" means to design something again or in different way (dictinary). Adding such sugar does not affect syntax, type checking, classes, constructions, the way you evaluate methods etc. It only affects global function resolution, which is worth considering. For me, consistency is always much better than having popups here and there, it just breaks your workflow. 

As I said earlier, I'm not sold to any of these ideas. I've also described the idea of removing self port as "arguably" good idea. However, for me, both proposals have some interesting advantages and disadvantages. 

Regarding errors - yeah, here I agree completely. If you make an error and you can auto-fix it , lets do it. However, the problem that its not obvious to what port connect node could be considered design flaw, not error (it could be considered error as well). It really needs discussion.

**Comment by kustosz** [comment]
Sorry, will check words in the dictionary before using :( :(
All I'm saying is that redesigning method/function dispatch is more of an RFC material, while this particular issue could turn into an awesome UX improvement that is valid regardless the syntax.

**Comment by mikusp** [comment]
Indeed it can be seen as two different issues: only suggesting conversion to method, no matter how, and providing a mechanism for applying common "fix-its".

If you think that it would be better to track them separately, please do so.

I created this issue cause it seems to be one the most asked questions, and without any feedback it's not clear why this code doesn't work. This could be also resolved by other means, I am not fixed on the one I proposed.

**Comment by wdanilo** [comment]
I agree! Having this form of auto-fixer for errors would be more than awesome. The question if using it for port connection is more related to the RFC's and is an open question to me now.
